apis/mapping/v1alpha1/namespaceoffloading_types.go

```diff
@@ -0,0 +1,96 @@
+/*
+
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	mapsv1alpha1 "github.com/liqotech/liqo/apis/virtualKubelet/v1alpha1"
+)
+
+type mappingStrategy string
+
+const (
+	// EqualName -> remote Namespace will have the same name of local Namespace.
+	EqualName mappingStrategy = "EnforceSame"
+	// StandardName -> remote Namespace will have a default name: localNamespaceName-localClusterID.
+	StandardName mappingStrategy = "Default"
+)
+
+type deployingStrategy string
+
+const (
+	// LocalDeploy allows user only to deploy pod on local cluster.
+	LocalDeploy deployingStrategy = "Local"
+	// RemoteDeploy allows user only to deploy pod on remote clusters, possibly filtered through ClusterSelector.
+	RemoteDeploy deployingStrategy = "Remote"
+	// LocalAndRemoteDeploy allows user to deploy pod locally and remotely.
+	LocalAndRemoteDeploy deployingStrategy = "LocalAndRemote"
+)
+
+// NamespaceOffloadingSpec defines the desired state of NamespaceOffloading.
+type NamespaceOffloadingSpec struct {
+	// RemappingNameStrategy allows user to map remote Namespace name with 2 different strategy
+	// (Look at mappingStrategy).
+	// +kubebuilder:validation:Enum="EnforceSame";"Default"
+	// +kubebuilder:default="Default"
+	RemappingNameStrategy mappingStrategy `json:"remappingNameStrategy,omitempty"`
+
+	// If DeployingPodStrategy allows user to deploy pods in three different scenario:
+	// { locally, remotely, locally and remotely}
+	// +kubebuilder:validation:Enum="Local";"Remote";"LocalAndRemote"
+	// +kubebuilder:default="LocalAndRemote"
+	DeployingPodStrategy deployingStrategy `json:"deployingPodStrategy,omitempty"`
+
+	// ClusterSelector allows user to specify the clusters distinguishing characteristics on which he wants to
+	// offload his pods. For every node label (key) is possible to specify a slice of values.
+	ClusterSelector map[string][]string `json:"clusterSelector,omitempty"`
+}
+
+// NamespaceOffloadingStatus defines the observed state of NamespaceOffloading.
+type NamespaceOffloadingStatus struct {
+	// Condition notifies user about different events happening during remote namespaces lifecycle.
+	Condition []corev1.NamespaceCondition `json:"condition,omitempty"`
```

Comment:
Should conditions be per remote cluster or global?

Reply:
I was thinking at gloabal conditions, but maybe is better to have these per remote cluster, what do you think about it ?

Reply:
My 2 cents: I also think it is much better to have per-cluster conditions. For instance, we can have offloading failing just on a specific cluster but working fine on the other ones.

Reply:
you are right, i will implement a single map with key (cluster-id) and value (struct with condition and phase)

Reply:
I agree per-cluster conditions would be better, as they provide more information. Though, I was wondering what it would happen in the tens/hundreds/thousands clusters scenario. In those cases, probably, it would be too much verbosity, but this optimization can be left to a later stage.

At the moment, I would add an overall Phase value, which summarizes the entire namespace offloading status and that can be shown as an additional column when getting the resource.